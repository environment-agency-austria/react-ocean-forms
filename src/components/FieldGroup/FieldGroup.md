### Examples

#### Simple field group

Showcase of a simple field group

```jsx
import { Form, FieldGroup, Input } from 'react-ocean-forms';

function Example() {
  // Submit callback, here you'd make your api calls
  const handleSubmit = (values) => {
    console.log('onSubmit, values: ' + JSON.stringify(values));
  };

  // Notice how a field group creates a sub object
  const defaultValues = {
    myGroup: {
      myInput1: 'input 1',
      myInput2: 'input 2',
    },
  };

  return (
    <Form className="demo" onSubmit={handleSubmit} defaultValues={defaultValues}>
      <FieldGroup
        name="myGroup"
        label="Example group"
        render={() => {
          return (
            <React.Fragment>
              <Input name="myInput1" label="Example input" />
              <Input name="myInput2" label="Example input 2" />
            </React.Fragment>
          );
        }}
      />
      <button type="submit">Submit</button>
    </Form>
  );
}

<Example />;
```
