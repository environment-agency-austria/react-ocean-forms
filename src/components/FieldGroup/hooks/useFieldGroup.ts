/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module FieldGroup
 */
import { useCallback, useMemo } from 'react';

import { useFormContext, useValidation, IValidationArgs, IBasicValidationState } from '../../../hooks';
import { useFullName, useFieldRegistration } from '../../../hooks/internal';
import { IFormContext, IFieldValues } from '../../FormContext';

import { getGroupValue } from './useFieldGroup.utils';
import { IUseFieldGroupArgs, IUseFieldGroupResult } from './useFieldGroup.types';

/**
 * @hidden
 */
export function useFieldGroup<TFieldValue = unknown>(props: IUseFieldGroupArgs<TFieldValue>): IUseFieldGroupResult {
  const formContext = useFormContext();

  const {
    name,
    label,
    defaultValues,
    values,
    disabled = formContext.disabled,
    plaintext = formContext.plaintext,
    asyncValidateOnChange = formContext.asyncValidateOnChange,
  } = props;

  const fullName = useFullName(name);
  const { validationState, validate, resetValidation, updateValidationState } = useValidation(props);

  /**
   * Triggers the validation of the group
   */
  const validateGroup = useCallback(
    (args?: Partial<IValidationArgs>): Promise<IBasicValidationState> => {
      const groupValue = getGroupValue<TFieldValue>(formContext, fullName);
      return validate(groupValue, args);
    },
    [formContext, fullName, validate],
  );

  // Register the group in the formContext, so the group
  // validation can be called on form submit.
  const registerFieldState = useMemo(() => ({
    label,
    isGroup: true,
    updateValidation: updateValidationState,
    validate: validateGroup,
    reset: resetValidation,
    getValue: () => ({}),
  }), [label, resetValidation, updateValidationState, validateGroup]);
  useFieldRegistration(
    fullName,
    registerFieldState,
  );

  /**
   * Listens to child field events, triggers validation if
   * needed and passes them to the higher context
   * @param name Field name
   * @param event Event name
   * @param args Event args
   */
  const notifyFieldEvent = useCallback(
    (name: string, event: string, args?: unknown): void => {
      formContext.notifyFieldEvent(name, event, args);

      if (event !== 'change' && event !== 'blur') { return; }

      if (event === 'change') {
        const localName = name.substring(fullName.length + 1);

        const currentGroupValue = getGroupValue<TFieldValue>(formContext, fullName);
        const intermediateGroupValue = {
          ...(currentGroupValue === undefined ? { } : currentGroupValue),
          ...{
            // Override the value of the event sender, because
            // the Field didn't update its state yet, making the
            // Form.getValues() returning an old Field value.
            [localName]: args,
          },
        };

        void validate(
          intermediateGroupValue as TFieldValue,
          { checkAsync: asyncValidateOnChange },
        );
      } else if (!asyncValidateOnChange) {
        void validate(getGroupValue(formContext, fullName));
      }
    },
    [asyncValidateOnChange, formContext, fullName, validate],
  );

  const subContext: IFormContext = useMemo(() => ({
    ...formContext,
    fieldPrefix: fullName,
    notifyFieldEvent,
    disabled,
    plaintext,
    defaultValues: (defaultValues === undefined ? formContext.defaultValues : { ...formContext.defaultValues, ...{ [fullName]: defaultValues }}) as Partial<IFieldValues>,
    values: (values === undefined ? formContext.values : { ...formContext.values, ...{ [fullName]: values }}) as Partial<IFieldValues>,
  }), [defaultValues, disabled, formContext, fullName, notifyFieldEvent, plaintext, values]);

  const groupState = useMemo(
    () => ({
      fullName,
      isValidating: validationState.isValidating,
      isRequired: validationState.isRequired,
      valid: validationState.valid,
      error: validationState.error,
    }),
    [
      fullName,
      validationState,
    ],
  );

  return {
    groupFormContext: subContext,
    renderParams: groupState,
  };
}
