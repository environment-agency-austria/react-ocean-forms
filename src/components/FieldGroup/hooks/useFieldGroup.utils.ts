/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @packageDocumentation
 * @module FieldGroup
 */
import { IFormContext } from '../../FormContext';

/**
 * Helper function to get the correct value
 * of the group (including all values of the nested fields)
 * @param formContext Form context
 * @param fullName Full name of the field group
 * @hidden
 */
export function getGroupValue<TFieldValue extends {}>(formContext: IFormContext, fullName: string): TFieldValue | undefined {
  const formValues = formContext.getValues();

  const formValue = formValues[fullName];
  if (formValue === '' || formValue === undefined) {
    return undefined;
  }

  return formValue as TFieldValue;
}
