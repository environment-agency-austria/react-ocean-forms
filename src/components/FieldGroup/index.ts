export * from './FieldGroup';
export * from './FieldGroup.types';
export { IFieldGroupRenderParams } from './hooks/useFieldGroup.types';
