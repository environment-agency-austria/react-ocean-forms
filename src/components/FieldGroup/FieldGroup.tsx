/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';

import { FormContext } from '../FormContext';
import { IFieldGroupProps } from './FieldGroup.types';
import { useFieldGroup } from './hooks/useFieldGroup';

/**
 * Wrapper for groups of input fields
 * managed by the form component
 */
export const FieldGroup: React.FC<IFieldGroupProps> = ({
  name, validators, asyncValidators, asyncValidationWait, render,
  disabled, plaintext, label, asyncValidateOnChange,
}): JSX.Element => {
  const [ subContext, groupState ] = useFieldGroup(
    name, label, validators, asyncValidators, asyncValidationWait,
    disabled, plaintext, asyncValidateOnChange,
  );

  return (
    <FormContext.Provider value={subContext}>
      {render(groupState)}
    </FormContext.Provider>
  );
}
