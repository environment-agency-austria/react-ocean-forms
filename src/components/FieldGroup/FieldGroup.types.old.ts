/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IValidationState } from '../../hooks';
import { IFieldValues } from '../FormContext';
import { IValidationProps } from '../withValidation';

/**
 * Meta information about the group
 * for the render prop
 */
export interface IFieldGroupRenderParams extends IValidationState {
  /**
   * Full name of the group
   */
  fullName: string;
}

/**
 * Props for the field group component
 */
export interface IFieldGroupProps<TFieldValues = IFieldValues> extends IValidationProps<TFieldValues> {
  /**
   * Field name
   */
  name: string;
  /**
   * Label of the group
   */
  label: string;
  /**
   * True, if the async validators should be triggered
   * on a change event
   */
  asyncValidateOnChange?: boolean;
  /**
   * Optional default values
   */
  defaultValues?: TFieldValues;
  /**
   * Optional values
   */
  values?: TFieldValues;
  /**
   * Disables this field group and all its fields.
   */
  disabled?: boolean;
  /**
   * Puts the field group and all its fields in plaintext mode.
   */
  plaintext?: boolean;
  /**
   * Render prop
   * @param params Meta information about the group
   */
  render(params: IFieldGroupRenderParams): JSX.Element;
}
