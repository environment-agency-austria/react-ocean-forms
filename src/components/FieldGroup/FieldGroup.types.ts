/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { IUseFieldGroupArgs, IFieldGroupRenderParams } from './hooks/useFieldGroup.types';

/**
 * Props for the field group component
 */
export interface IFieldGroupProps extends IUseFieldGroupArgs {
  /**
   * Gets called to render its children (see render prop pattern).
   * @param params Contains the group state consisting of fullName,
   * isValidating,valid, error which can be used to display those informations.
   */
  render(params: IFieldGroupRenderParams): JSX.Element;
}
