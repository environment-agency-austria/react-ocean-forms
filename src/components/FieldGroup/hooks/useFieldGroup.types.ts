import { IUseValidationArgs, IValidationState } from '../../../hooks';
import { IFormContext } from '../../FormContext';

/**
 * Props for the field group component
 */
export interface IUseFieldGroupArgs<TFieldValue = unknown> extends IUseValidationArgs<TFieldValue> {
  /**
   * Message id of the label that will be displayed along the input.
   * If you don't want to use any i18n features you can pass a raw message instead.
   */
  label: string;
  /**
   * If set to true the form will trigger asynchronous validation on Fields whenever
   * they change (e.g. on key press). Default behaviour is that the fields will only
   * async validate when they loose focus.
   * @default Form.asyncValidateOnChange
   */
  asyncValidateOnChange?: boolean;
  /**
   * Overwrites the Form default values for the child fields of this field group.
   * Those values will be put into the according fields when the form initializes.
   */
  defaultValues?: TFieldValue;
  /**
   * Overwrites the Form values for the child fields of this field group. Changing
   * this property will update all Field values, overwriting their default values
   * but also any value the user put in.
   */
  values?: TFieldValue;
  /**
   * Overwrites the disabled state for this field group.
   * @default Form.disabled
   */
  disabled?: boolean;
  /**
   * Overwrites the plaintext state for this field group.
   * @default Form.plaintext
   */
  plaintext?: boolean;
}

/**
 * Meta information about the group
 * for the render prop
 */
export interface IFieldGroupRenderParams extends IValidationState {
  /**
   * Full name of the group
   */
  fullName: string;
}

/**
 * Result of the useFieldGroup hook
 */
export interface IUseFieldGroupResult {
  /**
   * Overriden form context for the group
   */
  groupFormContext: IFormContext;
  /**
   * Params with meta information to be passed
   * to the render method of FieldGroup
   */
  renderParams: IFieldGroupRenderParams;
}
