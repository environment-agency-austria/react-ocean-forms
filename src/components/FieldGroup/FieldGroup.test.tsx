import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import {
  createMockFormContext,
  createMockValidationResult,
} from '../../test-utils/enzymeFormContext';
import { IFormContext } from '../FormContext';

import { FieldGroup } from './FieldGroup';
import { IFieldGroupProps } from './FieldGroup.types';
import { useFieldGroup } from './hooks/useFieldGroup';
import { IFieldGroupRenderParams } from './hooks/useFieldGroup.types';

jest.mock('./hooks/useFieldGroup');

describe('<FieldGroup />', () => {
  const mockName = 'unitGroup';
  const mockLabel = 'Unit group';

  interface ISetupArgs {
    props?: Partial<IFieldGroupProps<{} | undefined>>;
    contextOverrides?: Partial<IFormContext>;
    validationOverrides?: Partial<IFieldGroupRenderParams>;
  }

  interface ISetupResult {
    formContext: IFormContext;
    wrapper: ShallowWrapper;
    renderMock: jest.Mock;
    groupContext: IFormContext;
  }

  const setup = ({
    props,
    contextOverrides,
    validationOverrides,
  }: ISetupArgs = {}): ISetupResult => {
    const renderMock = jest.fn().mockReturnValue(null);

    const formContext: IFormContext = {
      ...createMockFormContext(),
      fieldPrefix: mockName,
      ...contextOverrides,
    };
    const { validationState } = createMockValidationResult();
    const validation: IFieldGroupRenderParams = {
      fullName: mockName,
      ...validationState,
      ...validationOverrides,
    };

    (useFieldGroup as jest.Mock).mockReturnValue({
      groupFormContext: formContext,
      renderParams: validation,
    });

    const wrapper = shallow(
      <FieldGroup name={mockName} label={mockLabel} render={renderMock} {...props} />
    );

    const groupContext = wrapper.first().prop('value') as IFormContext;

    return {
      formContext,
      wrapper,
      renderMock,
      groupContext,
    };
  };

  describe('Render', () => {
    it('should render without crashing', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Context overrides', () => {
    it('should create a valid form context', () => {
      const { formContext, groupContext } = setup();
      expect(groupContext).toMatchObject({
        fieldPrefix: mockName,
        registerField: formContext.registerField,
        unregisterField: formContext.unregisterField,
        notifyFieldEvent: formContext.notifyFieldEvent,
        plaintext: formContext.plaintext,
        registerListener: formContext.registerListener,
        unregisterListener: formContext.unregisterListener,
        getFieldState: formContext.getFieldState,
        getValues: formContext.getValues,
        busy: formContext.busy,
        disabled: formContext.disabled,
        asyncValidateOnChange: formContext.asyncValidateOnChange,
        asyncValidationWait: formContext.asyncValidationWait,
        defaultValues: formContext.defaultValues,
        stringFormatter: formContext.stringFormatter,
        submit: formContext.submit,
      });
    });
  });

  describe('render prop', () => {
    it('should get called with the correct parameters', () => {
      const { renderMock } = setup();
      expect(renderMock).toHaveBeenCalledWith({
        fullName: mockName,
        isValidating: false,
        isRequired: false,
        valid: true,
        error: null,
      });
    });

    it('should correctly pass the validation state', () => {
      const mockIsValidating = true;
      const mockIsRequired = true;
      const mockIsValid = false;
      const mockError = { message_id: 'bar', params: {} };

      const { renderMock } = setup({
        validationOverrides: {
          isValidating: mockIsValidating,
          isRequired: mockIsRequired,
          valid: mockIsValid,
          error: mockError,
        },
      });
      expect(renderMock).toHaveBeenCalledWith({
        fullName: mockName,
        isValidating: mockIsValidating,
        isRequired: mockIsRequired,
        valid: mockIsValid,
        error: mockError,
      });
    });
  });
});
