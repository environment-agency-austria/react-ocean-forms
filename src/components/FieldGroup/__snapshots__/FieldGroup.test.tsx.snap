// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldGroup /> Render should render without crashing 1`] = `
<ContextProvider
  value={
    Object {
      "asyncValidateOnChange": false,
      "asyncValidationWait": 400,
      "busy": false,
      "defaultValues": Object {},
      "disabled": false,
      "fieldPrefix": "unitGroup",
      "getFieldState": [MockFunction],
      "getValues": [MockFunction],
      "notifyFieldEvent": [MockFunction],
      "plaintext": false,
      "registerField": [MockFunction],
      "registerListener": [MockFunction],
      "stringFormatter": [MockFunction],
      "submit": [MockFunction],
      "unregisterField": [MockFunction],
      "unregisterListener": [MockFunction],
    }
  }
/>
`;
